### PR TITLE
Move ChannelConfig static fields to ChannelHandshakeConfig

### PR DESCRIPTION
**Pull request by wpaulino**
In the near future, we plan to allow users to update their `ChannelConfig` after the initial channel handshake. In order to reuse the same struct and expose it to users, we opt to move out all static fields that cannot be updated after the initial channel handshake.

`announced_channel` and `commit_upfront_shutdown_pubkey` represent these fields. We now store `announced_channel` within the `Channel` struct, making sure we're backwards compatible to allow users to downgrade if necessary. `commit_upfront_shutdown_pubkey` doesn't need to be stored since it only determines whether a
`Channel` stores a `shutdown_scriptpubkey` or not prior to a `Channel` being stored.

Motivated by https://github.com/lightningdevkit/rust-lightning/pull/1527#discussion_r892512704.

### CODE REVIEW

File: lightning/src/util/config.rs [hunk 1]
```diff
@@ -265,18 +276,8 @@ pub struct ChannelConfig {
...
-	/// This should only be set to true for nodes which expect to be online reliably.
-	///
-	/// As the node which funds a channel picks this value this will only apply for new outbound
-	/// channels unless [`ChannelHandshakeLimits::force_announced_channel_preference`] is set.
-	///
-	/// This cannot be changed after the initial channel handshake.
-	///
-	/// Default value: false.
-	pub announced_channel: bool,
+	/// Deprecated, use [`ChannelHandshakeConfig::announced_channel`] instead.
```
Comment (tnull) [inline lightning/src/util/config.rs:294]: Hm, if you deprecate this here, could you update the rest of the codebase to directly use the new location?

Also, instead of a comment, you probably want to mark this as deprecated:
```
#[deprecated(
	since = "0.0.107",
	note = "ChannelHandshakeConfig::announced_channel should be used instead.",
)]
```
(Yes, if you cannot update all instances where it is used immediately, this will produce annoying warning messages, which is however exactly the idea.)

File: lightning/src/util/config.rs [hunk 2]
```diff
@@ -278,18 +293,8 @@ pub struct ChannelConfig {
...
-	///
-	/// This is set to true by default to provide a slight increase in security, though ultimately
-	/// any attacker who is able to take control of a channel can just as easily send the funds via
-	/// lightning payments, so we never require that our counterparties support this option.
-	///
-	/// This cannot be changed after a channel has been initialized.
-	///
-	/// Default value: true.
-	pub commit_upfront_shutdown_pubkey: bool,
+	/// Deprecated, use [`ChannelHandshakeConfig::commit_upfront_shutdown_pubkey`] instead.
```
Comment (tnull) [inline lightning/src/util/config.rs:296]: Same here, you probably want to use:

```
#[deprecated(
	since = "0.0.107",
	note = "ChannelHandshakeConfig::commit_upfront_shutdown_pubkey should be used instead.",
)]
```

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -928,9 +929,17 @@ impl<Signer: Sign> Channel<Signer> {
 			}
 		}
 
+		// Also update `ChannelConfig::announced_channel` to make sure we still allow users to
+		// downgrade versions after the fact.
+		//
+		// TODO: This can be removed once we prevent downgrades to v0.0.107.
```
Comment (tnull) [inline lightning/src/ln/channel.rs:935]: Not 100% sure what exactly you mean here with "prevent downgrades to v0.0.107". Do you basically mean "when we're ready to break backwards compat with versions prior to v0.0.107"?

File: lightning/src/ln/channel.rs [hunk 4]
```diff
@@ -495,6 +495,7 @@ pub(super) struct Channel<Signer: Sign> {
 	#[cfg(not(any(test, feature = "_test_utils")))]
 	config: ChannelConfig,
 
+	announced: bool,
```
Comment (tnull) [inline lightning/src/ln/channel.rs:498]: Why is this third instance of an `announced` bool (besides the new `ChannelHandshakeConfig::announced_channel` and the now-deprecated `ChannelConfig::announced_channel`) necessary? Since it seems to be instantiated with `own_channel_config.announced_channel`, couldn't we just use this or move it entirely to `Channel`?

**Review by tnull — COMMENTED** [review]
Some comments and questions.

Moreover: Not sure if its only me, but I find the current naming of variables in `channel.rs` vs. their type unnecessarily confusing: there are many instances of variables simply named `config`, which may refer to `ChannelConfig`, `UserConfig`, etc. depending on the context. Then there is the `channel_options` field and corresponding variables, which is however a `ChannelConfig`. Moreover, there are fields like `own_channel_config`, which would make you think it would be a `ChannelConfig`, but no, it's actually a `ChannelHandshakeConfig`.

TLDR: Not sure if it needs to happen in this PR, but I personally would appreciate a refactoring that brings the naming of variables more in alignment with their types to improve readability.

File: lightning/src/util/config.rs (hunk 1 above)
Comment (tnull) [reply lightning/src/util/config.rs:294]: Mh, I agree that in that timeframe it may be come a nuisance without much benefit. Maybe we could create something like a `TODO.md` in which we keep track of TODOs that cannot be completed right away and where we define until when they are deferred? Or immediately create issues for them with a 'defer' tag and assign an estimated milestone some time in the future?

Comment (wpaulino) [reply lightning/src/util/config.rs:294]: > In any case, I think we can do better here. Instead of making these fields pub(crate), because we don't really care about users being able to serialize ChannelConfig objects, I think, we can move the serialization to some struct LegacyChannelConfig { config: ChannelConfig, announced_channel: bool, ... } object instead. That way, we can read it all the same, but the public interface remains clean and you can construct a full ChannelConfig without using default() and then modifying. Its a good bit more work cause we can't use the impl_writeable_tlv_based!() macro anymore and have to write out the deserialization manually, but it should be doable.

Sounds good, I assume this means we can reset the serialization schema for the new `ChannelConfig` then? Otherwise, we would want to document that the types 4 (`announced_channel`) & 6 (`commit_upfront_shutdown_key`) cannot be reused as it would break `LegacyChannelConfig` serialization.

> Maybe we could create something like a TODO.md in which we keep track of TODOs that cannot be completed right away and where we define until when they are deferred? Or immediately create issues for them with a 'defer' tag and assign an estimated milestone some time in the future?

:+1: I would prefer the latter.

File: lightning/src/ln/channel.rs (hunk 3 above)
Comment (wpaulino) [reply lightning/src/ln/channel.rs:935]: Correct, I can reword to your suggestion if it's not clear as is.

File: lightning/src/ln/channel.rs (hunk 4 above)
Comment (wpaulino) [reply lightning/src/ln/channel.rs:498]: `ChannelHandshakeConfig` doesn't need to be stored anywhere (and currently isn't) as it's mostly ephemeral, but the `announced` bit should definitely continue to be stored, so I opted to have it within the `Channel` struct.

File: lightning/src/util/config.rs [hunk 5]
```diff
@@ -265,30 +291,6 @@ pub struct ChannelConfig {
 	///
 	/// [`MIN_CLTV_EXPIRY_DELTA`]: crate::ln::channelmanager::MIN_CLTV_EXPIRY_DELTA
 	pub cltv_expiry_delta: u16,
-	/// Set to announce the channel publicly and notify all nodes that they can route via this
```
Comment (tnull) [inline lightning/src/util/config.rs:268]: nit: If you split up the commits, you may want to move the removal of the moved parts to the 'move' commits. (Nevermind, if you end up squashing it all).

File: lightning/src/util/config.rs [hunk 6]
```diff
@@ -337,23 +339,83 @@ impl Default for ChannelConfig {
...
-	(4, announced_channel, required),
-	(6, commit_upfront_shutdown_pubkey, required),
-	(8, forwarding_fee_base_msat, required),
-});
+/// Legacy version of [`ChannelConfig`] that stored the static
+/// [`ChannelHandshakeConfig::announced_channel`] and
+/// [`ChannelHandshakeConfig::commit_upfront_shutdown_pubkey`] fields.
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct LegacyChannelConfig {
+	pub(crate) mutable: ChannelConfig,
```
Comment (tnull) [inline lightning/src/util/config.rs:353]: Not too sure about this field name. Is mutability its core characteristic/function?

**Review by tnull — APPROVED** [review]
LGTM, just two remarks, feel free to ignore.
